@benesch Do you know what is going on with the teamcity failure:

```
[16:31:59][Build archive] set -euo pipefail; \
[16:31:59][Build archive]   ret=$(cd ./pkg/sql/parser/gen && goyacc -p sql -o sql.go.tmp sql.y); \
[16:31:59][Build archive]   if expr "$ret" : ".*conflicts" >/dev/null; then \
[16:31:59][Build archive]     echo "$ret"; exit 1; \
[16:31:59][Build archive]   fi
[16:31:59][Build archive] (echo "// Code generated by make. DO NOT EDIT."; \
[16:31:59][Build archive]  echo "// GENERATED FILE DO NOT EDIT"; \
[16:31:59][Build archive]  echo; \
[16:31:59][Build archive]  echo "package lex"; \
[16:31:59][Build archive]  echo; \
[16:31:59][Build archive]  grep '^const [A-Z][_A-Z0-9]* ' pkg/sql/parser/gen/sql.go.tmp) > pkg/sql/lex/tokens.go
[16:31:59][Build archive] (echo "// Code generated by goyacc. DO NOT EDIT."; \
[16:31:59][Build archive]  echo "// GENERATED FILE DO NOT EDIT"; \
[16:31:59][Build archive]  cat pkg/sql/parser/gen/sql.go.tmp | \
[16:31:59][Build archive]  sed -E 's/^const ([A-Z][_A-Z0-9]*) =.*$/const \1 = lex.\1/g') > pkg/sql/parser/sql.go
[16:32:05][Build archive] build/node-run.sh -C ./pkg/ui yarn install
[16:32:05][Build archive] error An unexpected error occurred: "https://registry.yarnpkg.com/svgo/-/svgo-0.7.2.tgz: Request failed \"500 Internal Server Error\"".
[16:32:05][Build archive] yarn install v1.5.1
[16:32:05][Build archive] make: *** [pkg/ui/yarn.installed] Error 1
[16:32:05][Build archive] [1/5] Validating package.json...
[16:32:05][Build archive] [2/5] Resolving packages...
[16:32:05][Build archive] [3/5] Fetching packages...
[16:32:05][Build archive] info If you think this is a bug, please open a bug report with the information provided in "/go/src/github.com/cockroachdb/cockroach/pkg/ui/yarn-error.log".
[16:32:05][Build archive] info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
[16:32:05][Build archive] Makefile:322: recipe for target 'pkg/ui/yarn.installed' failed
[16:32:06][Build archive] Process exited with code 2
```

--- a/pkg/storage/engine/rocksdb.go
+++ b/pkg/storage/engine/rocksdb.go
@@ -627,7 +627,6 @@ func (r *RocksDB) open() error {
 func (r *RocksDB) syncLoop() {
 	s := &r.syncer
 	s.Lock()
-	defer s.Unlock()
 
 	var lastSync time.Time
 
@@ -636,6 +635,7 @@ func (r *RocksDB) syncLoop() {
 			s.cond.Wait()
 		}
 		if s.closed {
+			s.Unlock()
 			return
 		}
 
@@ -1692,7 +1692,7 @@ func (r *rocksDBBatch) Commit(syncCommit bool) error {
 	if leader {
 		// We're the leader of our group. Wait for any running commit to finish and
 		// for our batch to make it to the head of the pending queue.
-		for c.committing && c.pending[0] != r {
+		for c.committing || c.pending[0] != r {
 			c.cond.Wait()
 		}
 
@@ -1728,14 +1728,19 @@ func (r *rocksDBBatch) Commit(syncCommit bool) error {
 		// proceed.
 		c.Lock()
 		c.committing = false
-		c.cond.Signal()
+		// NB: Multiple leaders can be waiting.
+		c.cond.Broadcast()
 		c.Unlock()
 
 		// Propagate the error to all of the batches involved in the commit. If a
 		// batch requires syncing and the commit was successful, add it to the
 		// syncing list. Note that we're reusing the pending list here for the
-		// syncing list.
-		syncing := pending[:0]
+		// syncing list. We need to be careful to cap the capacity so that
+		// extending this slice past the length of the pending list will result in
+		// reallocation. Otherwise we have a race between appending to this list
+		// while holding the sync lock below, and appending to the commit pending
+		// list while holding the commit lock above.
+		syncing := pending[:0:len(pending)]
 		for _, b := range pending {
 			if err != nil || !b.syncCommit {
 				b.commitErr = err

